spec/aptible/cli/helpers/s3_log_helpers_spec.rb

```diff
@@ -0,0 +1,276 @@
+        Aws::S3::Resource.new(region: 'us-east-1', client: client_stub)
+      end
+    end
+
+    it 'finds files with a single matching string' do
+      strings = %w(yoda)
+      result = subject.find_s3_files_by_string_match('us-east-1', 'bucket',
+                                                     'stack', strings)
+      expect(result).to match_array(%w(stack/matter/it/does/not/yoda))
+    end
+
+    it 'finds files with two matching strings' do
+      strings = %w(it matter)
+      result = subject.find_s3_files_by_string_match('us-east-1', 'bucket',
+                                                     'stack', strings)
+      expect(result).to match_array(%w(stack/it/doesnt/matter
+                                       stack/matter/it/does/not/yoda))
+    end
+
+    it 'only find files with all matching strings' do
+      strings = %w(it yoda)
+      result = subject.find_s3_files_by_string_match('us-east-1', 'bucket',
+                                                     'stack', strings)
+      expect(result).to match_array(%w(stack/matter/it/does/not/yoda))
+    end
+  end
+
+  describe '#find_s3_files_by_attrs' do
+    before do
+      client_stub = Aws::S3::Client.new(stub_responses: true)
+      client_stub.stub_responses(
+        :list_buckets, buckets: [{ name: 'bucket' }]
+      )
+      client_stub.stub_responses(
+        :list_objects_v2, contents: [
+          { key: v2app },
+          { key: v2app_rotated },
+          { key: v3db_rotated },
+          { key: v3db },
+          { key: v3app }
+        ]
+      )
+      subject.stub(:s3_client) do
+        Aws::S3::Resource.new(region: 'us-east-1', client: client_stub)
+      end
+    end
+
+    it 'can find apps by id' do
+      attrs = { type: 'apps', id: 321 }
+      result = subject.find_s3_files_by_attrs('us-east-1', 'bucket',
+                                              'stack', attrs)
+      expect(result).to match_array([v3app, v2app, v2app_rotated])
+    end
+
+    it 'can find databases by id' do
+      attrs = { type: 'databases', id: 321 }
+      result = subject.find_s3_files_by_attrs('us-east-1', 'bucket',
+                                              'stack', attrs)
+      expect(result).to match_array([v3db, v3db_rotated])
+    end
+
+    it 'can find by other attributes of the log file like container id' do
+      attrs = { container_id: 'deadbeef' }
+      result = subject.find_s3_files_by_attrs('us-east-1', 'bucket',
+                                              'stack', attrs)
+      expect(result).to match_array([v3app])
+    end
+  end
+
+  describe '#time_match?' do
+    let(:first_log) { Time.parse('2022-08-01T00:00:00') }
+    let(:last_log) { Time.parse('2022-09-01T00:00:00') }
+    let(:before) { Time.parse('2022-07-01T00:00:00') }
+    let(:between) { Time.parse('2022-08-15T00:00:00') }
+    let(:after) { Time.parse('2022-10-01T00:00:00') }
+
+    context 'identifies files that may have lines within a range' do
```

Comment:
i would include your ascii diagram comment here and correspond each to a case, the one you shared:

```
  |  |lu
  | l|u
 l|  |u
  |lu|
 l|u |
lu|  |

l = lower bound of search
u= upper bound of search
|'s are the first and last timestamp in the file
```

and mention each on a line here as it was very helpful

Reply:
Will do!